bin/zkServer.sh

```diff
@@ -235,51 +235,64 @@ restart)
+      if ! [[ "$myid" =~ ^[0-9]+$ ]] ; then
+        echo "myid could not be determined, will not able to locate clientPort in the server configs."
+      else
+        clientPortAndAddress=`$GREP "^[[:space:]]*server.$myid=.*;.*" "$ZOOCFG" | sed -e 's/.*=//' | sed -e 's/.*;//'`
+        if [ ! "$clientPortAndAddress" ] ; then
+          echo "Client port not found in static config file. Looking in dynamic config file."
+          dynamicConfigFile=`$GREP "^[[:space:]]*dynamicConfigFile" "$ZOOCFG" | sed -e 's/.*=//'`
+          clientPortAndAddress=`$GREP "^[[:space:]]*server.$myid=.*;.*" "$dynamicConfigFile" | sed -e 's/.*=//' | sed -e 's/.*;//'`
+        fi
+        if [ ! "$clientPortAndAddress" ] ; then
+          echo "Client port not found in the server configs"
+        else
+          if [[ "$clientPortAndAddress" =~ ^.*:[0-9]+ ]] ; then
+            clientPortAddress=`echo "$clientPortAndAddress" | sed -e 's/:.*//'`
+          fi
+          clientPort=`echo "$clientPortAndAddress" | sed -e 's/.*://'`
+        fi
+      fi
     fi
-    echo "Client port found: $clientPort. Client address: $clientPortAddress."
+    if [ ! "$clientPort" ] ; then
+      echo "Client port not found. Looking for secureClientPort in the static config."
+      secureClientPort=`$GREP "^[[:space:]]*secureClientPort[^[:alpha:]]" "$ZOOCFG" | sed -e 's/.*=//'`
+      if [ "$secureClientPort" ] ; then
+        isSSL="true"
+      fi
+      clientPort=$secureClientPort
+    fi
+    if [ ! "$clientPort" ] ; then
+      echo "Unable to find either secure or unsecure client port in any configs. Terminating."
+      exit 1
+    fi
+    echo "Client port found: $clientPort. Client address: $clientPortAddress. Client SSL: $isSSL."
     STAT=`"$JAVA" "-Dzookeeper.log.dir=${ZOO_LOG_DIR}" "-Dzookeeper.root.logger=${ZOO_LOG4J_PROP}" "-Dzookeeper.log.file=${ZOO_LOG_FILE}" \
-             -cp "$CLASSPATH" $JVMFLAGS org.apache.zookeeper.client.FourLetterWordMain \
-             $clientPortAddress $clientPort srvr 2> /dev/null    \
+          -cp "$CLASSPATH" $CLIENT_JVMFLAGS $JVMFLAGS org.apache.zookeeper.client.FourLetterWordMain \
+          $clientPortAddress $clientPort srvr $isSSL 2> /dev/null    \
           | $GREP Mode`
     if [ "x$STAT" = "x" ]
     then
-        echo "Error contacting service. It is probably not running."
-        exit 1
+      if [ "$isSSL" ] ; then
+        echo "We used secureClientPort ($secureClientPort) to establish connection, but we failed. Please make sure you provided all the Client SSL connection related parameters in the SERVER_JVMFLAGS environment variable!"
+        echo "example: SERVER_JVMFLAGS=\"-Dzookeeper.clientCnxnSocket=org.apache.zookeeper.ClientCnxnSocketNetty -Dzookeeper.ssl.trustStore.location=/tmp/clienttrust.jks -Dzookeeper.ssl.trustStore.password=password -Dzookeeper.ssl.keyStore.location=/tmp/client.jks -Dzookeeper.ssl.keyStore.password=password -Dzookeeper.client.secure=true\" ./zkServer.sh status"
```

Comment:
Why do we need "Client SSL connection related parameters in the SERVER_JVMFLAGS"? 
(Missing the concept from my end)

Reply:
yep, this is not very intuitive.
- the status command in zkServer.sh will initiate a client connection to execute diagnostic command against the server. To be able to connect to the server, it needs to have all the client-side credentials / ssl configs (and these configs can not be read from the server-side zoo.cfg file)
- normally you put these client SSL parameters to the CLIENT_JVMFLAGS environment variable. But the zkServer.sh extra configuration goes usually to the SERVER_JVMFLAGS. We could change this by processing the CLIENT_JVMFLAGS during the `zkServer.sh status` command, but I am not sure if it would be less confusing.

Reply:
I will add a bit more context to the documentation / error message.

Reply:
Thinking it through again, I think it will be actually less confusing to use the CLIENT_JVMFLAGS. I fixed it in my last commit. Thanks for the note!

Reply:
Thank you for explaining the logic :) 
If we need the flags set, can we do an if check in the zkEnv.sh if its TLS enabled or not? So when zkCli runs, it will get the CLIENT_JVMFLAGS which already has everything needed. As running zkServer with all the flags might be a bit tedious.

Reply:
Also, do you know if running the 4 letter commands are SSL secured or not?

Reply:
4lw commands are not ssl secured, you can't use it with SSL port.
You need to provide those flags if you want to connect to a secure port.

Reply:
"can we do an if check in the zkEnv.sh" - Sounds good, but I'm not sure we want to complicate zkServer.sh with client jvm flag checks. I'll leave this for Mate to decide, he is much more involved in this.

Reply:
> "can we do an if check in the zkEnv.sh"

I was also thinking about more automation, but I don't think it would be easy to do this reliably. The problem is that usually (in production environment) we use different keystore / truststore files (or jaas configs for SASL) for the server and for the client. And in the `zoo.conf` file or in  `zkEnv.sh` we don't have the SSL config needed for the client to connect. Usually it is something one configures in the code or by using system properties when one tries to connect to ZooKeeper.

But maybe I misunderstood your idea...

Reply:
apologize for the delay in reverting. I didn't know that to enable client-side TLS changes we always need to set in the JVM flags. I thought setting them in the zoo.cfg (using ssl.keyStore.location, ssl.keyStore.password, ssl.keyStore.location and ssl.keyStore.password ) will help to achieve the same feat. So I suggested if we can set the keystore and trustore related config by default in an SSL enabled env. But I see your point regarding dev and prod having different paths for either of the stores. 
But we can define some ENV variables to check if its prod or dev and still set the appropriate configs in zkEnv.sh right? Like just providing flexibility so that people don't have to put it everytime they do a status check (or maybe I am overthinking all this :) )

Reply:
@nkalmar Thank you for testing this out. Regarding 4lw, we use netcat like,
**echo ruok | nc <zookeeper ip> 2181**
so if we pass the CLIENT_JVMFLAGS will it be SSL protected?

Reply:
> But we can define some ENV variables to check if its prod or dev and still set the appropriate configs in zkEnv.sh right? Like just providing flexibility so that people don't have to put it everytime they do a status check (or maybe I am overthinking all this :) )

I think anyone can easily make an alias or a one-liner shell script for his environment if he needs to call the zkServer.sh with some extra environment variables, so I wouldn't complicate this any more :)

Reply:
> so if we pass the CLIENT_JVMFLAGS will it be SSL protected?

no, the CLIENT_JVMFLAGS will not be used / processed by `nc`. I think you will have to specify the key / trust files with some netcat specific options. (see `nc --help`)